core/src/test/java/org/apache/iceberg/TestPartitionSpecUpdate.java

```diff
@@ -0,0 +1,255 @@
+  public void testAddSamePartitionField() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+
+    if (formatVersion == 1) {
+      table.updateSpec()
+          .addBucketField("data", 16, "data_partition")
+          .commit();
+      Assert.assertEquals("[\n" +
+          "  1000: data_bucket: bucket[16](2)\n" +
+          "  1001: data_partition: bucket[16](2)\n" +
+          "]", table.spec().toString());
+    } else {
+      AssertHelpers.assertThrows(
+          "Should throw IllegalArgumentException if adding a duplicate partition field",
+          IllegalArgumentException.class,
+          "Field Id 1000 has already been used in the existing partition fields",
+          () -> table.updateSpec()
+              .addBucketField("data", 16, "data_partition")
+              .commit());
+    }
+  }
+
+  @Test
+  public void testAddField() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1000, table.spec().lastAssignedFieldId());
+
+    table.updateSpec()
+        .addBucketField("data", 8, "data_partition")
+        .addBucketField("id", 8)
+        .addBucketField("data", 6, "data_field")
+        .commit();
+
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "  1001: data_partition: bucket[8](2)\n" +
+        "  1002: id_bucket: bucket[8](1)\n" +
+        "  1003: data_field: bucket[6](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1003, table.spec().lastAssignedFieldId());
+  }
+
+  @Test
+  public void testRenameField() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1000, table.spec().lastAssignedFieldId());
+    Assert.assertEquals(0, table.spec().specId());
+
+    table.updateSpec()
+        .renameField("data_bucket", "data_partition")
+        .addBucketField("id", 8)
+        .commit();
+
+    Assert.assertEquals("[\n" +
+        "  1000: data_partition: bucket[16](2)\n" +
+        "  1001: id_bucket: bucket[8](1)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1001, table.spec().lastAssignedFieldId());
+    Assert.assertEquals(1, table.spec().specId());
+  }
+
+  @Test
+  public void testRenameFieldExceptions() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1000, table.spec().lastAssignedFieldId());
+    Assert.assertEquals(0, table.spec().specId());
+
+    AssertHelpers.assertThrows(
+        "Should throw IllegalArgumentException if only renaming a partition field",
+        IllegalArgumentException.class,
+        "Cannot set default partition spec to the current default",
+        () -> table.updateSpec()
+            .renameField("data_bucket", "data_partition")
```

Comment:
I think this should be allowed. We should just rebuild the spec with a different name.